[analyzer] Add option assume-at-least-one-iteration

NagyDonat opened the pull request:
This commit adds the new analyzer option `assume-at-least-one-iteration`, which is `false` by default, but can be set to `true` to ensure that the analyzer always assumes at least one iteration in loops.

In some situations this "loop is skipped" execution path is an important corner case that may evade the notice of the developer and hide significant bugs -- however, there are also many situations where it's guaranteed that at least one iteration will happen (e.g. some data structure is always nonempty), but the analyzer cannot realize this and will produce false positives when it assumes that the loop is skipped.

This commit refactors some logic around the implementation of the new feature, but the only functional change is introducing the new analyzer option. If the new option is left in its default state (false), then the analysis is functionally equivalent to an analysis done with a version before this commit.

-----

This commit is the new step of my [loop handling improvement plans](https://discourse.llvm.org/t/loop-handling-improvement-plans/80417). When I wrote this, I hoped (based on eyeballing older analysis run results) that this will be an important "must have" improvement of the analyzer results; however it turns out that the effects are more ambiguous.

Based on this, I no longer think that this option should be enabled by default, but I still think that this option could be useful for some users to reduce the amount of unwanted results.

I analyzed the usual set of open source projects to see the effects of enabling this option, and I started to evaluate the results, but I didn't finish this evaluation because I felt that even this partial evaluation is enough to justify merging this as an off-by-default feature.

| Project | Assuming one iteration | Baseline | Notes |
|---------|------------------------|----------| ----- |
| memcached | [0 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=memcached_1.6.8_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=memcached_1.6.8_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [1 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=memcached_1.6.8_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=memcached_1.6.8_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) | assuming empty list -> calloc with zero size report |
| tmux | [0 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=tmux_2.6_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=tmux_2.6_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [7 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=tmux_2.6_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=tmux_2.6_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) | all 7 are core.NonNullParamChecker reports that assume emptiness of custom data structures, e.g. [this](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=tmux_2.6_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=tmux_2.6_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved&report-id=6875004&report-hash=db2dd8ecbd79514cf57950e9cdc116e0&report-filepath=tmux%2Fwindow-buffer.c) |
| curl | [0 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=curl_curl-7_66_0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=curl_curl-7_66_0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [0 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=curl_curl-7_66_0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=curl_curl-7_66_0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) | |
| twin | [0 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [3 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) | one [corner case](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved&report-id=6876057&report-hash=ec682267df95619957c1f3658029ca16&report-filepath=twin%2Fserver%2Frcrun.c) that seems to be very unlikely but might be a true positive, one [false positive](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved&report-id=6876104&report-hash=a71674c6b490eb0e63d8fd26c1fb7cb5&report-filepath=twin%2Flibs%2FlibTw%2FlibTw.c) assumes an empty string, one [unrelated false positive](https://codechecker-demo.eastus.cloudapp.azure.com/Default/report-detail?run=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=twin_v0.8.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved&report-id=6876107&report-hash=b922b94f0857d8c03c7a7f76d38e717d&report-filepath=twin%2Flibs%2FlibTw%2FlibTw.c) accidentally discarded
| vim | [5 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=vim_v8.2.1920_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=vim_v8.2.1920_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [18 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=vim_v8.2.1920_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=vim_v8.2.1920_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) | 
| openssl | [0 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=openssl_openssl-3.0.0-alpha7_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=openssl_openssl-3.0.0-alpha7_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [5 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=openssl_openssl-3.0.0-alpha7_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=openssl_openssl-3.0.0-alpha7_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| sqlite | [12 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=sqlite_version-3.33.0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=sqlite_version-3.33.0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [25 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=sqlite_version-3.33.0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=sqlite_version-3.33.0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| ffmpeg | [24 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=ffmpeg_n4.3.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=ffmpeg_n4.3.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [175 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=ffmpeg_n4.3.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=ffmpeg_n4.3.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| postgres | [2 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=postgres_REL_13_0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=postgres_REL_13_0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [49 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=postgres_REL_13_0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=postgres_REL_13_0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| tinyxml2 | [0 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=tinyxml2_8.0.0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=tinyxml2_8.0.0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [0 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=tinyxml2_8.0.0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=tinyxml2_8.0.0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| libwebm | [3 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=libwebm_libwebm-1.0.0.27_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=libwebm_libwebm-1.0.0.27_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [0 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=libwebm_libwebm-1.0.0.27_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=libwebm_libwebm-1.0.0.27_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| xerces | [1 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=xerces_v3.2.3_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=xerces_v3.2.3_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [3 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=xerces_v3.2.3_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=xerces_v3.2.3_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| bitcoin | [0 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=bitcoin_v0.20.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=bitcoin_v0.20.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [9 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=bitcoin_v0.20.1_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=bitcoin_v0.20.1_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| protobuf | [0 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=protobuf_v3.13.0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=protobuf_v3.13.0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [4 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=protobuf_v3.13.0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=protobuf_v3.13.0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved) 
| qtbase | [16 new reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=qtbase_v6.2.0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=qtbase_v6.2.0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=New) | [49 resolved reports](https://codechecker-demo.eastus.cloudapp.azure.com/Default/reports?run=qtbase_v6.2.0_edonnag_Ericsson-assume-first-iteration_a4c7718&newcheck=qtbase_v6.2.0_edonnag_Ericsson-assume-first-iteration_b3c6379&diff-type=Resolved)

NagyDonat committed:
[analyzer][NFC] Add option assume-one-iteration

This commit adds the new analyzer option `assume-one-iteration`, which
is `false` by default, but can be set to `true` to ensure that the
analyzer always assumes (at least) one iteration in loops.

In some situations this "loop is skipped" execution path is an
important corner case that may evade the notice of the developer and
hide significant bugs -- however, there are also many situations where
it's guaranteed that at least one iteration will happen (e.g. some
data structure is always nonempty), but the analyzer cannot realize
this and will produce false positives when it assumes that the loop is
skipped.

This commit refactors some logic around the implementation of the new
feature, but those changes are clearly NFC when the new analyzer option
is in its default state.

NagyDonat commented:
By the way is the "NFC" tag justified for this "add an off-by-default option" commit? I used this because e.g. if somebody is looking for a bug, they can safely skip this commit -- but OTOH this _does_ add a new feature.

Xazax-hun commented:
My preference would be to not have NFC tags for off by default features.

steakhal commented on clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:
```diff
@@ -294,6 +294,16 @@ ANALYZER_OPTION(
     "Whether the analysis should try to unroll loops with known bounds.", false)
 
+ANALYZER_OPTION(
+    bool, ShouldAssumeOneIteration, "assume-one-iteration",
+    "Whether the analyzer should always assume at least one iteration in "
```
I wish the name would reflect that it's **at least one**, and not exactly one iteration.

NagyDonat replied:
I agree, but I cannot think of a name that reflects that without being awkwardly long. Perhaps "may-assume-skipped-loop" (with a negated meaning i.e. the default is true)?

steakhal replied:
How about the naive "assume-at-least-one-iteration"?
"assume" already implies the part that the condition was opaque, otherwise we wouldn't assume anything there.

NagyDonat replied:
> How about the naive "assume-at-least-one-iteration"?

I felt that it's a bit too long, but if you prefer it (and other reviewers aren't opposed), then I can accept it.

steakhal replied:
Many other options are even longer.
I prefer this.

whisperity replied:
`assume-loops-iterate`? (As opposed to "not iterating", i.e., running 0 iterations.)

NagyDonat replied:
> `assume-loops-iterate`? (As opposed to "not iterating", i.e., running 0 iterations.)

Nope, this would be very confusing if I didn't already understand the meaning of this feature.

steakhal commented:
Overall I had a quick look and nothing sticks out, but the devil is in the details, right?
So I plan to allocate some quality time to review this PR this week.

Thanks for the PR!

NagyDonat commented:
> My preference would be to not have NFC tags for off by default features.

OK, I removed the [NFC] tag from the PR title and adjusted the description accordingly.

gamesh411 commented on clang/include/clang/StaticAnalyzer/Core/AnalyzerOptions.def:
```diff
@@ -294,6 +294,16 @@ ANALYZER_OPTION(
+    "determine if it's true or false). Setting this to true eliminates some "
+    "false positives (where e.g. a structure is nonempty, but the analyzer "
+    "does not notice this); but it also eliminates some true positives (e.g. "
+    "cases where a structure can be empty and this causes buggy behavior).",
```
It could be mentined that this option interacts with analyzer option 'eagerly-assume'.
We could just say somthing the enabling both can lead to more paths explored around loops, or just leave it at that, and suggest that the user should be avare of the connection between the two.
IMO either one is fine, because if someone is interested enough to read this option text, they can either do their measurements tweaking their parameters, or looking at the history and seeing the test file for more details.
What do you think?

NagyDonat replied:
> It could be mentined that this option interacts with analyzer option 'eagerly-assume'.

It's just a bug -- not an undocumented feature :upside_down_face: . 

The _implementation_ of my change needs to heavily interact with `eagerly-assume` but I'm trying to reimplement the same behavior with and without it. There are some contrived corner cases where enabling or disabling `eagerly-assume` changes the behavior of this `assume-one-iteration` logic, but I wasn't able to construct an example that could plausibly appear in real-world code, so I don't think that we need to document these theoretical buggy cases.

gamesh411 replied:
That seems fair; it is good to go.

gamesh411 reviewed (commented):
This enables a heuristic that is closer to real-world code usage, and having this option is valuable. I left one comment inline.

steakhal reviewed (approved):
Reviewed the changes and it looks correct.
Please change the flag name (along with the comments to `assume-at-least-one-iteration`.

Please also review Edre's comments, those appear relevant.

NagyDonat committed:
Rename option to assume-at-least-one-iteration

NagyDonat committed:
Clarify comments about relationship with eagerly-assume

NagyDonat commented:
I updated the name of the option and tried to clarify the comments that describe the relationship with `eagerly-assume` (but it's just a mostly-theoretical buggy corner case, so I wouldn't put it into the minimal two-sentence documentation).

@gamesh411 @steakhal Are you satisfied with this?
